hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -101,29 +128,113 @@ public void shutdown(boolean terminate) {
+        getOrCreateRingbufferContainers(partitionId).put(ringbuffer.getNamespace(), ringbuffer);
+    }
+
     @Override
     public void beforeMigration(PartitionMigrationEvent partitionMigrationEvent) {
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Here we will replicate only ringbuffer containers for the ringbuffer service. Other services are responsible
+     * for migrating their ringbuffer containers on migration events. The main reason is because of fine-grained
+     * anti entropy (replication). The partition data might not be migrated in full but rather in part.
+     * For instance, the partition data for a single map might not be in sync between replicas but the ringbuffer
+     * data might be in sync. This means that the map data will be replicated but the ringbuffer data will not be replicated
+     * which can cause differences between ringbuffer data for some services.
+     * In another words, each service is responsible for replicating all of it's data, including data in other services.
+     *
+     * @param event replication event
+     * @return the replication operation
+     */
     @Override
     public Operation prepareReplicationOperation(PartitionReplicationEvent event) {
-        Map<String, RingbufferContainer> migrationData = new HashMap<String, RingbufferContainer>();
-        IPartitionService partitionService = nodeEngine.getPartitionService();
-        for (Map.Entry<String, RingbufferContainer> entry : containers.entrySet()) {
-            String name = entry.getKey();
-            int partitionId = partitionService.getPartitionId(getPartitionKey(name));
-            RingbufferContainer container = entry.getValue();
-            int backupCount = container.getConfig().getTotalBackupCount();
-            if (partitionId == event.getPartitionId() && backupCount >= event.getReplicaIndex()) {
-                migrationData.put(name, container);
+        return prepareReplicationOperation(event, getAllServiceNamespaces(event));
+    }
+
+    @Override
+    public Operation prepareReplicationOperation(PartitionReplicationEvent event,
+                                                 Collection<ServiceNamespace> namespaces) {
+        final int partitionId = event.getPartitionId();
+        final Map<ObjectNamespace, RingbufferContainer> partitionContainers = containers.get(partitionId);
+        if (partitionContainers == null || partitionContainers.isEmpty()) {
+            return null;
+        }
+        final Map<String, RingbufferContainer> migrationData = new HashMap<String, RingbufferContainer>();
+        for (ServiceNamespace namespace : namespaces) {
+            final ObjectNamespace ns = (ObjectNamespace) namespace;
+            final RingbufferContainer container = partitionContainers.get(ns);
+            if (container != null && container.getConfig().getTotalBackupCount() >= event.getReplicaIndex()) {
+                migrationData.put(ns.getObjectName(), container);
```

Comment:
How is `ns.getObjectName()` going to be converted a full namespace on target, without having service name?